src/interfaces/chain.h

```diff
@@ -245,8 +228,38 @@ class Chain
         virtual void chainStateFlushed(const CBlockLocator& locator) {}
     };
 
-    //! Register handler for notifications.
-    virtual std::unique_ptr<Handler> handleNotifications(std::shared_ptr<Notifications> notifications) = 0;
+    using ScanFn = std::function<Optional<uint256>(const uint256& start_hash, int start_height, const uint256& tip_hash, int tip_height)>;
+    using MempoolFn = std::function<void(std::vector<CTransactionRef>)>;
+
+    //! Register handler for notifications. Before returning and before sending
+    //! the first notification, call ScanFn with a scan range of blocks after a
+    //! specified location and time, and then call MempoolFn with a snapshot of
+    //! transactions in the mempool before the first transactionAddedToMempool /
+    //! TransactionRemovedFromMempool notifications.
+    //!
+    //! @param[in] notify        callback object receiving notifications
+    //! @param[in] scan_fn       callback invoked before notifications are sent
+    //!                          to scan blocks after a specified location and
+    //!                          time. This should return the hash of the last
+    //!                          block scanned, and may be called more than once
+    //!                          if new blocks were connected during the scan.
+    //! @param[in] mempool_fn    callback invoked before notifications are sent
+    //!                          with snapshot of mempool transactions
+    //! @param[in] scan_locator  location of last block previously scanned.
+    //!                          scan_fn will be only be called for blocks after
+    //!                          this point. Can be null to scan from genesis.
+    //! @param[in] scan_time     minimum block timestamp for beginning the scan
+    //!                          scan_fn will only be called for blocks starting
+    //!                          from this timestamp
+    //! @param[out] tip          information about chain tip at the point where
+    //!                          notifications will begin
+    virtual std::unique_ptr<Handler> handleNotifications(std::shared_ptr<Notifications> notifications,
```

Comment:
I think it would be better to rename `handleNotifications`. IMO it does 3 different tasks:
* trigger rescan
* sync with mempool
* register notification

I'm fine with doing all of three in same functions rather than 3 chunks for each, but name could be something like `syncClient` ?

Also you may be interested by rescan but not mempool-sync and therefore it may be made optional.